### PR TITLE
allow for 31 occupied bits in vanilla `BoxedSmallWord`

### LINKED ISSUE #3327: `if_tagged_scalar` only checks LSBit, so `box`ing of `Nat32` should shift by 31 bits
Currently we heap allocate too much. We should `get_n; shr 31; if ...` but we compare against `1 << 30`.

This was correct when the key bit vas the second LSBit, but we changed this a long time ago (the crucial commit was 365b4b8d471b14c18e5c525c757ec7135c1bc171).

Even better: we could `get_n; clz; if ...`

The situation is not completely easy. `BoxedSmallWord` covers both `Nat32` and `Int32`, and (see comment below) this trick can only handle `Nat32`. So in order to enable the more efficient tagging scheme for `Nat32` (which is used by hashed containers all  over the base library) we have to  introduce a new stack representation.

### PR DESCRIPTION
**Pull request by ggreif**
fixes #3327

- [ ] Needs a bunch more tests, since the bitwise-operations may also be affected w.r.t bit 31 in `Int32`.

### CODE REVIEW

File: src/codegen/compile.ml [hunk 1]
```diff
@@ -6550,15 +6598,18 @@ module StackRep = struct
...
+    | UnboxedUnsigned32, Vanilla -> BoxedSmallWord.Unsigned.box env
+    | Vanilla, UnboxedUnsigned32 -> BoxedSmallWord.Unsigned.unbox env
 
     | UnboxedFloat64, Vanilla -> Float.box env
     | Vanilla, UnboxedFloat64 -> Float.unbox env
 
     | Const (_, Const.Lit (Const.Bool b)), Vanilla -> Bool.lit b
     | Const c, Vanilla -> compile_unboxed_const (materialize_const_t env c)
-    | Const (_, Const.Lit (Const.Word32 n)), UnboxedWord32 -> compile_unboxed_const n
+    | Const (_, Const.Lit (Const.Word32 n)), UnboxedWord32 -> compile_unboxed_const n (* FIXME: what about UnboxedUnsigned32? *)
```
Comment (ggreif) [inline src/codegen/compile.ml:6612]: This looks like an elephant in this room. Maybe we need `Const.Unsigned32`?

Comment (ggreif) [reply src/codegen/compile.ml:6612]: Yes this looks doable.

File: src/codegen/compile.ml [hunk 2]
```diff
@@ -6550,15 +6598,18 @@ module StackRep = struct
     | UnboxedWord64, Vanilla -> BoxedWord64.box env
     | Vanilla, UnboxedWord64 -> BoxedWord64.unbox env
 
-    | UnboxedWord32, Vanilla -> BoxedSmallWord.box env
-    | Vanilla, UnboxedWord32 -> BoxedSmallWord.unbox env
+    | UnboxedWord32, Vanilla -> BoxedSmallWord.boxI env
+    | Vanilla, UnboxedWord32 -> BoxedSmallWord.unboxI env
+
+    | UnboxedUnsigned32, Vanilla -> BoxedSmallWord.Unsigned.box env
+    | Vanilla, UnboxedUnsigned32 -> BoxedSmallWord.Unsigned.unbox env
```
[the remark this answers is not in the document]
Comment (crusso) [reply src/codegen/compile.ml:6605]: What's the difference between UnboxUnsigned32 and UnboxedWord32? Trapping vs Wrapped?

[the remark this answers is not in the document]
Comment (nomeata) [reply src/codegen/compile.ml:6605]: Can't we use the same boxing/unboxing strategy? I.e. treat them not as numbers, but bitstrings?

At the level of the backend, I think it makes sense to have only one unboxed representation for 32bit numbers, and only operations care about signedness.

At least I think that may be easier to think about. And matches what wasm does.

[the remark this answers is not in the document]
Comment (ggreif) [reply src/codegen/compile.ml:6605]: Let's go on a tangent about an efficient `00`/`11` check... What about
``` wasm
get n
shr 30      # 4 possibilities, `00`/`11`: good, `01`/`10`: bad
popcnt      # 3 possibilities: `0`, `2`: good `1`: bad
and 1      # 2 possibilities, `0`: good, `1`: bad
if ...
```
This is not even that bad :-)

[the remark this answers is not in the document]
Comment (ggreif) [reply src/codegen/compile.ml:6605]: Contrast this with the `Nat32` optimal boxing sequence:
``` wasm
get n
clz         # 2 possibilities, `0`: bad, else: good
if ...
```
3 instructions

[the remark this answers is not in the document]
Comment (nomeata) [reply src/codegen/compile.ml:6605]: And if you just use the 0 check for both? It means negative int32 always get boxed, but maybe that's ok?

[the remark this answers is not in the document]
Comment (ggreif) [reply src/codegen/compile.ml:6605]: The 3-instruction test is not a good choice for `Int32`, as all negative numbers end up on the heap. This is why separate boxing strategies would be optimal.

**Comment by ggreif** [comment]
Closing this one, since #3335 gives a similar bang for much less buck. We can reheat this if we absolutely want.

**Pull request closed without merging**